### PR TITLE
Add public preview API in DataMap

### PR DESCRIPTION
**Pull request by yifan-zhou922**
# Add public preview APIs for

- Audit log API
This audit API has been previously reviewed. 
Issue: https://github.com/Azure/azure-rest-api-specs/issues/24004
PR: https://github.com/Azure/azure-rest-api-specs/pull/24133
Path: POST /audit/query
Updates:
   -  Using continuationToken instead of watermark

- Entity delete with hierarchy
This API has been previously reviewed.
PR: https://github.com/Azure/azure-rest-api-specs/pull/23095
Issue: https://github.com/Azure/azure-rest-api-specs/issues/23248
Updates: 
    - Using query parameter parallelCount
   - Instead of using hierarchy within the api path, we use entity/bulk/delete
 Path:
DELETE /entity/bulk/delete
GET /entity/operations/{operationId}

**Diff:** https://openapihub.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL2dpdGh1Yi5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvYmxvYi9tYWluL3NwZWNpZmljYXRpb24vcHVydmlldy9kYXRhLXBsYW5lL0F6dXJlLkFuYWx5dGljcy5QdXJ2aWV3LkRhdGFNYXAvc3RhYmxlLzIwMjMtMDktMDEvcHVydmlld2RhdGFtYXAuanNvbiIsInRhcmdldFVybCI6Imh0dHBzOi8vZ2l0aHViLmNvbS9BenVyZS9henVyZS1yZXN0LWFwaS1zcGVjcy9ibG9iL3lpZmFuemhvdTkyMi0xMC0wMS1wcmV2aWV3L3NwZWNpZmljYXRpb24vcHVydmlldy9kYXRhLXBsYW5lL0F6dXJlLkFuYWx5dGljcy5QdXJ2aWV3LkRhdGFNYXAvcHJldmlldy8yMDIzLTEwLTAxLXByZXZpZXcvcHVydmlld2RhdGFtYXAuanNvbiJ9

<a href="?expand=1&template=data_plane_template.md">Click here</a> to open a PR for a Data Plane API.

<a href="?expand=1&template=control_plane_template.md">Click here</a> to open a PR for a Control Plane (ARM) API.

### CODE REVIEW

**Comment by yifan-zhou922** [comment]
Hi @mikekistler , @MushMal , could you please help review the API changes for this public preview version?  These two set of APIs have been go through the review meeting and updated accordingly. Many thanks.

**Comment by yifan-zhou922** [comment]
The new version is missing a client parameter that was found in the old version. Was 'operationGuid' removed or renamed?
New: [Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json#L7271:3](https://github.com/Azure/azure-rest-api-specs/blob/f1f7b44d16682fafb289583ddba9bb2cdcf51704/specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json#L7271:3)
Old: [Azure.Analytics.Purview.DataMap/stable/2023-09-01/purviewdatamap.json#L6779:3](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/purviewdatamap.json#L6779:3)

For this breaking changes, the original operationGuid was never used in the old version. In the new version, we rename it to operationId to align with the azure guidelines.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (weidongxu-microsoft) [inline specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1649]: ??? what's about csv?

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (weidongxu-microsoft) [inline specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1685]: ?

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: Typo. Fixed.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: Typo. Fixed.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: Thanks. Changed to EntityLongRunningOpeartion

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: Thanks. Fixed.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1660]: The cancel operation will return the job metadata similar to get operation. However, the status will be set to Aborted. cc @EldonZhao 
{
    "id": "44a12e08-c4ce-469f-8d6f-246ac077959d",
    "status": "Aborted",
    "createTime": "2023-12-19T03:02:50.378+0000",
    "taskPickUpTime": "2023-12-19T03:03:17.472+0000",
    "lastUpdateTime": "2023-12-19T03:06:02.029+0000",
    "properties": {
    },
    "error": {
        "code": "",
        "message": ""
    },
    "createdBy": "ServiceAdmin",
    "progress": {
        "status": null,
        "statistics": null
    }
}

Comment (weidongxu-microsoft) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1660]: It is highly recommended (near required) that the terminal status is one of `Succeeded, Failed, or Canceled`
https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md#long-running-operations

One part of the story, SDK relied on this 3 values as the end of LRO polling.

A minor part compared to above, it seems hard for user to know how much of the bulk is done, in such case? (I assume it is not atomic)

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (TomDu) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:4148]: Yes it's for pagination.
The API invokes M365 Audit API internally. And the layout is the way that M365 audit does pagination. We followed the same contract. If it's not a strong requirement, we may keep it to have a consistent experience.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1593]: Thanks. it is only for bulk delete API. Description udpated.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: Thanks for pointing this out. Updated to operation-location.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1660]: Thanks for information. We plan to exclude this cancel operation in this api version release. We will do the API changes according to the guidance.

Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1660]: Whether PartialSucceeded will be treated as the terminal state?

Comment (weidongxu-microsoft) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: No, it is not (healthinsights asked the same question).

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (weidongxu-microsoft) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:1593]: Sure. This may get back from SDK review, that a `ParallelCount` (if you get to typespec) would be too generic.

So far this is fine for Swagger and preview.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (weidongxu-microsoft) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:4112]: "M365 Audit API" appears to be your implementation detail? Does customer know what they are working with M365 Audit API? Or large part of your API is just passthrough from M365 Audit API?

If not, seems not a good reason to diverge. Again, exception happens, but required a good reason (e.g. Atlas, or all pageable in your service is done this way).

Comment (TomDu) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:4112]: Thank you Weidong. Today Purview Audit is based on M365 audit solution. Users today can directly view M365 audit logs via M365 unified audit portal and M365 audit query API, so we think we can provide a consistent experience for the user.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: Hi @weidongxu-microsoft , we have updated to operation-location according to this doc. https://github.com/Azure/autorest/blob/main/docs/extensions/readme.md#x-ms-long-running-operation-options

However, when we tried to create a diff doc, the error was raised.
ERROR: Schema violation: No enum match for: operation-location 

https://openapihub.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL2dpdGh1Yi5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvYmxvYi9tYWluL3NwZWNpZmljYXRpb24vcHVydmlldy9kYXRhLXBsYW5lL0F6dXJlLkFuYWx5dGljcy5QdXJ2aWV3LkRhdGFNYXAvc3RhYmxlLzIwMjMtMDktMDEvcHVydmlld2RhdGFtYXAuanNvbiIsInRhcmdldFVybCI6Imh0dHBzOi8vZ2l0aHViLmNvbS9BenVyZS9henVyZS1yZXN0LWFwaS1zcGVjcy9ibG9iL3lpZmFuemhvdTkyMi0xMC0wMS1wcmV2aWV3L3NwZWNpZmljYXRpb24vcHVydmlldy9kYXRhLXBsYW5lL0F6dXJlLkFuYWx5dGljcy5QdXJ2aWV3LkRhdGFNYXAvcHJldmlldy8yMDIzLTEwLTAxLXByZXZpZXcvcHVydmlld2RhdGFtYXAuanNvbiJ9

Comment (weidongxu-microsoft) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json]: Emm, maybe a bug on diff. Doc says explicitly that it is one of the options.

If you want to workaround the diff error, you maybe able to remove the `x-ms-long-running-operation-options`. It is usually not required (Swagger converted from TypeSpec seems not generate it currently e.g https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-07-01-preview/generated.json#L264) Without it at least better than a wrong one.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (weidongxu-microsoft) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:4112]: I cannot say I buy in that. But at preview this probably not a blocking issue.

When you go GA this API, please make sure we discuss this in review meeting.

**Comment by yifan-zhou922** [comment]
Thanks @weidongxu-microsoft  for the review comments. Could you please help signoff this preview version changes if there are no other comments needed for discussion. Many thanks.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (weidongxu-microsoft) [inline specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:7200]: Why is it not modeled? It is hard for customer to use without any modeling of the detail.

Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:7200]: The value will be different for different long running entity operations. The properties here should be a Map<String,String> where the keys will not be the same for different long running operations.

**Comment by weidongxu-microsoft** [comment]
API need further review
- https://github.com/Azure/azure-rest-api-specs/pull/26689#discussion_r1427468482
- https://github.com/Azure/azure-rest-api-specs/pull/26689#discussion_r1432314451

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (yifan-zhou922) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:7200]: Thanks. We will also fix this after we fully onboard with TypeSpec.

File: specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json
Comment (weidongxu-microsoft) [reply specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json:4112]: SDK may require you do customized code to make this works like a pageable.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by ms-henglu** [comment]
Please get a breaking changes approval from Azure Breaking Changes Reviewers <azbreakchangereview@microsoft.com> and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

**Comment by weidongxu-microsoft** [comment]
> The new version is missing a client parameter that was found in the old version. Was 'operationGuid' removed or renamed? New: [Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json#L7271:3](https://github.com/Azure/azure-rest-api-specs/blob/f1f7b44d16682fafb289583ddba9bb2cdcf51704/specification/purview/data-plane/Azure.Analytics.Purview.DataMap/preview/2023-10-01-preview/purviewdatamap.json#L7271:3) Old: [Azure.Analytics.Purview.DataMap/stable/2023-09-01/purviewdatamap.json#L6779:3](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/purview/data-plane/Azure.Analytics.Purview.DataMap/stable/2023-09-01/purviewdatamap.json#L6779:3)
> 
> For this breaking changes, the original operationGuid was never used in the old version. In the new version, we rename it to operationId to align with the azure guidelines.

Skip the breaking change CI. As detailed by Yifan, the removed model was never used anywhere. So removing should not affect anyone (no SDK released for DataMap).

This is PR for data-plane. I will ignore ARM labels.

**Comment by weidongxu-microsoft** [comment]
> API need further review/discussion
> 
> * [Add public preview API in DataMap #26689 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/26689#discussion_r1427468482)
> * [Add public preview API in DataMap #26689 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/26689#discussion_r1432314451)

One again, do review these in next preview or before GA

**Pull request merged**
